pyop2/mpi.py

```diff
@@ -183,46 +409,58 @@ def free_comm(comm, remove=True):
     This only actually calls MPI_Comm_free once the refcount drops to
     zero.
     """
-    if comm == MPI.COMM_NULL:
-        return
-    refcount = comm.Get_attr(refcount_keyval)
-    if refcount is None:
-        # Not a PyOP2 communicator, check for an embedded comm.
-        comm = comm.Get_attr(innercomm_keyval)
-        if comm is None:
-            raise ValueError("Trying to destroy communicator not known to PyOP2")
-        refcount = comm.Get_attr(refcount_keyval)
-        if refcount is None:
-            raise ValueError("Inner comm without a refcount")
-
-    refcount[0] -= 1
-
-    if refcount[0] == 0:
+    # ~ if isinstance(comm, list):
+    # ~ import pytest; pytest.set_trace()
+    if comm != MPI.COMM_NULL:
+        assert is_pyop2_comm(comm)
         ocomm = comm.Get_attr(outercomm_keyval)
+        if isinstance(ocomm, list):
+            # No idea why this happens!?
+            raise ValueError("Why have we got a list!?")
```

Comment:
:D

Reply:
This is hopefully fixed :crossed_fingers: